File: games/welcome_to/tables.py
"""The static-table signature -- ``RUST_PORT_PLAN.md`` M0-D.

Both engines hash the rule tables they were built from and compare at load.  A
silent table divergence (one deck count off, one plan score transposed, a codec
offset shifted) produces a legal-looking game that is a *different game*, and it
would pass every per-action gate: the two engines would agree perfectly about a
world neither of them shares with BGA.

⚠ **The signature is over integers, not text.** Every table is flattened to a
stream of ``i64`` and folded with FNV-1a; nothing depends on formatting,
encoding or float repr, all three of which are places Rust and Python may
legitimately differ.  Strings appearing in plan parameters are mapped through
:data:`_PARAM_CODES`, so an unrecognised one raises rather than hashing to
whatever ``str`` happens to produce.

Change a table on purpose and this number changes; that is the point.  Update
the expected value in ``tests/test_tables.py``, and rebuild the Rust crate --
the mismatch is a build reminder, not a bug.
"""

from __future__ import annotations

from typing import Iterator

from games.welcome_to import action_codec as codec
from games.welcome_to.constants import (
    BIS_BOXES,
    BIS_SCORES,
    BOX_OFFSET,
    CARD_TABLE,
    ESTATE_ROW_BOXES,
    ESTATE_ROW_SCORES,
    EXTREMITY_POSITIONS,
    FENCE_OFFSET,
    FENCE_SIZES,
    MAX_ESTATE_SIZE,
    MAX_NUMBER,
    MIN_NUMBER,
    NUM_BOXES,
    NUM_FENCES,
    NUM_STREETS,
    PARK_BOXES,
    PARK_SCORES,
    PERMIT_BOXES,
    PERMIT_SCORES,
    POOL_BOXES,
    POOL_POSITIONS,
    POOL_SCORES,
    ROUNDABOUT,
    ROUNDABOUT_BOXES,
    ROUNDABOUT_SCORES,
    SOLO_CARD_ID,
    SOLO_DECK_MIDDLE,
    STREET_SIZES,
    TEMP_BOXES,
    TEMP_DELTAS,
    TEMP_RANK_SCORES,
    TEMP_SOLO_SCORE,
    TEMP_SOLO_THRESHOLD,
)
from games.welcome_to.plans import PLANS

#: Bump when the *stream layout* changes (not when a table's values change).
SIGNATURE_VERSION: int = 1

#: Plan parameters that are strings.  Fixed codes, so the stream stays integral.
_PARAM_CODES: dict[str, int] = {
    "park": 1,
    "pool": 2,
    "pool&park": 3,
    "iceCream": 4,
    "without": 5,
    "with": 6,
    "christmas": 7,
    "easterEgg": 8,
}

#: The codec sections, in layout order.  Named explicitly rather than iterating
#: ``OFFSET``, so that a section renamed or reordered is a signature change.
_CODEC_SECTIONS: tuple[str, ...] = (
    "CHOOSE_STACK",
    "PERMIT_REFUSAL",
    "ROUNDABOUT_OPEN",
    "ROUNDABOUT_POS",
    "WRITE",
    "SURVEYOR_FENCE",
    "ESTATE_ROW",
    "PARK_STREET",
    "POOL_BUILD",
    "BIS",
    "CHOOSE_PLAN",
    "VALIDATE_ESTATE",
    "PASS_ROUNDABOUT",
    "PASS_SURVEYOR",
    "PASS_ESTATE",
    "PASS_PARK",
    "PASS_POOL",
    "PASS_BIS",
    "PASS_PLAN",
    "RESHUFFLE_YES",
    "RESHUFFLE_NO",
)

_FNV_OFFSET = 0xCBF29CE484222325
_FNV_PRIME = 0x100000001B3
_MASK64 = (1 << 64) - 1


def _param_value(param) -> int:
    if isinstance(param, bool):  # bool is an int; nothing uses one, so refuse
        raise TypeError("plan parameters are ints or known strings")
    if isinstance(param, int):
        return param
    try:
        return _PARAM_CODES[param]
    except KeyError:
        raise ValueError(
            f"plan parameter {param!r} has no signature code; add it to "
            "_PARAM_CODES in both engines"
        ) from None


def signature_stream() -> Iterator[int]:
    """Every static table, flattened, in a fixed order.  Mirrored in Rust."""
    yield SIGNATURE_VERSION

    # -- cards ------------------------------------------------------------
    yield len(CARD_TABLE)
    for number, effect in CARD_TABLE:
        yield -1 if number is None else number
        yield int(effect)
    yield SOLO_CARD_ID
    yield SOLO_DECK_MIDDLE

    # -- geometry ---------------------------------------------------------
    yield NUM_STREETS
    yield from STREET_SIZES
    yield from FENCE_SIZES
    yield from BOX_OFFSET
    yield from FENCE_OFFSET
    yield NUM_BOXES
    yield NUM_FENCES
    yield MIN_NUMBER
    yield MAX_NUMBER
    yield ROUNDABOUT
    yield from TEMP_DELTAS

    # -- score tracks -----------------------------------------------------
    yield from PARK_BOXES
    for row in PARK_SCORES:
        yield len(row)
        yield from row
    yield POOL_BOXES
    yield from POOL_SCORES
    for x, y in POOL_POSITIONS:
        yield x
        yield y
    yield BIS_BOXES
    yield from BIS_SCORES
    yield TEMP_BOXES
    yield from TEMP_RANK_SCORES
    yield TEMP_SOLO_THRESHOLD
    yield TEMP_SOLO_SCORE
    yield MAX_ESTATE_SIZE
    yield from ESTATE_ROW_BOXES
    for row in ESTATE_ROW_SCORES:
        yield len(row)
        yield from row
    yield PERMIT_BOXES
    yield from PERMIT_SCORES
    yield ROUNDABOUT_BOXES
    yield from ROUNDABOUT_SCORES
    for x, y in EXTREMITY_POSITIONS:
        yield x
        yield y

    # -- plans ------------------------------------------------------------
    yield len(PLANS)
    for plan in PLANS:
        yield plan.id
        yield int(plan.variant)
        yield plan.stack
        yield plan.scores[0]
        yield plan.scores[1]
        yield int(plan.kind)
        yield len(plan.params)
        for param in plan.params:
            yield _param_value(param)

    # -- codec layout -----------------------------------------------------
    yield codec.NUM_ACTIONS
    for name in _CODEC_SECTIONS:
        yield codec.OFFSET[name]
    for i, j in codec.EXPERT_PAIRS:
        yield i
        yield j


def table_signature() -> int:
    """FNV-1a 64 over :func:`signature_stream`, each value as 8 little-endian
    bytes.  ``welcome_to_rust::tables::table_signature`` computes the same.

    FNV rather than SHA-256 deliberately: the crate's dependency list starts at
    pyo3 only (``RUST_PORT_PLAN.md`` §6), and this hash guards against a
    transcription slip, not an adversary."""
    h = _FNV_OFFSET
    for value in signature_stream():
        word = value & _MASK64
        for shift in range(0, 64, 8):
            h ^= (word >> shift) & 0xFF
            h = (h * _FNV_PRIME) & _MASK64
    return h


if __name__ == "__main__":  # pragma: no cover - a convenience for the Rust side
    print(f"0x{table_signature():016x}")

File: games/welcome_to/welcome_to_rust/src/tables.rs
//! The static-table signature — a mirror of `games/welcome_to/tables.py`
//! (RUST_PORT_PLAN.md M0-D).
//!
//! Both engines hash the rule tables they were built from and compare at load.
//! A silent table divergence produces a legal-looking game that is a *different
//! game*, and it would pass every per-action gate: the two engines would agree
//! perfectly about a world neither shares with BGA.
//!
//! ⚠ The stream is integers, not text: nothing here depends on formatting,
//! encoding or float repr, all three of which Rust and Python may legitimately
//! disagree about.

use crate::codec;
use crate::constants::{
    card_table, solo_card_id, BIS_BOXES, BIS_SCORES, BOX_OFFSET, ESTATE_ROW_BOXES,
    ESTATE_ROW_SCORES, EXTREMITY_POSITIONS, FENCE_OFFSET, FENCE_SIZES, MAX_ESTATE_SIZE,
    MAX_NUMBER, MIN_NUMBER, NUM_BOXES, NUM_FENCES, NUM_STREETS, PARK_BOXES, PARK_SCORES,
    PERMIT_BOXES, PERMIT_SCORES, POOL_BOXES, POOL_POSITIONS, POOL_SCORES, ROUNDABOUT,
    ROUNDABOUT_BOXES, ROUNDABOUT_SCORES, SOLO_DECK_MIDDLE, STREET_SIZES, TEMP_BOXES,
    TEMP_DELTAS, TEMP_RANK_SCORES, TEMP_SOLO_SCORE, TEMP_SOLO_THRESHOLD,
};
use crate::plans::{Param, PLANS};

/// Bump when the *stream layout* changes (not when a table's values change).
pub const SIGNATURE_VERSION: i64 = 1;

const FNV_OFFSET: u64 = 0xCBF2_9CE4_8422_2325;
const FNV_PRIME: u64 = 0x0000_0100_0000_01B3;

/// Fixed codes for the plan parameters that are strings.
fn param_value(param: Param) -> i64 {
    match param {
        Param::Int(v) => v as i64,
        Param::Text(text) => match text {
            "park" => 1,
            "pool" => 2,
            "pool&park" => 3,
            "iceCream" => 4,
            "without" => 5,
            "with" => 6,
            "christmas" => 7,
            "easterEgg" => 8,
            other => panic!(
                "plan parameter {other:?} has no signature code; add it to both engines"
            ),
        },
    }
}

/// Every static table, flattened, in the order `tables.py::signature_stream`
/// yields them.
pub fn signature_stream() -> Vec<i64> {
    let mut out: Vec<i64> = Vec::with_capacity(768);
    out.push(SIGNATURE_VERSION);

    // -- cards ------------------------------------------------------------
    out.push(card_table().len() as i64);
    for &(number, effect) in card_table().iter() {
        out.push(number as i64);
        out.push(effect.as_i64());
    }
    out.push(solo_card_id() as i64);
    out.push(SOLO_DECK_MIDDLE as i64);

    // -- geometry ---------------------------------------------------------
    out.push(NUM_STREETS as i64);
    out.extend(STREET_SIZES.iter().map(|&v| v as i64));
    out.extend(FENCE_SIZES.iter().map(|&v| v as i64));
    out.extend(BOX_OFFSET.iter().map(|&v| v as i64));
    out.extend(FENCE_OFFSET.iter().map(|&v| v as i64));
    out.push(NUM_BOXES as i64);
    out.push(NUM_FENCES as i64);
    out.push(MIN_NUMBER as i64);
    out.push(MAX_NUMBER as i64);
    out.push(ROUNDABOUT as i64);
    out.extend(TEMP_DELTAS.iter().map(|&v| v as i64));

    // -- score tracks -----------------------------------------------------
    out.extend(PARK_BOXES.iter().map(|&v| v as i64));
    for row in PARK_SCORES.iter() {
        out.push(row.len() as i64);
        out.extend(row.iter().map(|&v| v as i64));
    }
    out.push(POOL_BOXES as i64);
    out.extend(POOL_SCORES.iter().map(|&v| v as i64));
    for &(x, y) in POOL_POSITIONS.iter() {
        out.push(x as i64);
        out.push(y as i64);
    }
    out.push(BIS_BOXES as i64);
    out.extend(BIS_SCORES.iter().map(|&v| v as i64));
    out.push(TEMP_BOXES as i64);
    out.extend(TEMP_RANK_SCORES.iter().map(|&v| v as i64));
    out.push(TEMP_SOLO_THRESHOLD as i64);
    out.push(TEMP_SOLO_SCORE as i64);
    out.push(MAX_ESTATE_SIZE as i64);
    out.extend(ESTATE_ROW_BOXES.iter().map(|&v| v as i64));
    for row in ESTATE_ROW_SCORES.iter() {
        out.push(row.len() as i64);
        out.extend(row.iter().map(|&v| v as i64));
    }
    out.push(PERMIT_BOXES as i64);
    out.extend(PERMIT_SCORES.iter().map(|&v| v as i64));
    out.push(ROUNDABOUT_BOXES as i64);
    out.extend(ROUNDABOUT_SCORES.iter().map(|&v| v as i64));
    for &(x, y) in EXTREMITY_POSITIONS.iter() {
        out.push(x as i64);
        out.push(y as i64);
    }

    // -- plans ------------------------------------------------------------
    out.push(PLANS.len() as i64);
    for plan in PLANS.iter() {
        out.push(plan.id as i64);
        out.push(plan.variant as u8 as i64);
        out.push(plan.stack as i64);
        out.push(plan.scores.0 as i64);
        out.push(plan.scores.1 as i64);
        out.push(plan.kind as u8 as i64);
        out.push(plan.params.len() as i64);
        for &param in plan.params.iter() {
            out.push(param_value(param));
        }
    }

    // -- codec layout -----------------------------------------------------
    out.push(codec::NUM_ACTIONS as i64);
    for offset in [
        codec::A_CHOOSE_STACK,
        codec::A_PERMIT_REFUSAL,
        codec::A_ROUNDABOUT_OPEN,
        codec::A_ROUNDABOUT_POS,
        codec::A_WRITE,
        codec::A_SURVEYOR_FENCE,
        codec::A_ESTATE_ROW,
        codec::A_PARK_STREET,
        codec::A_POOL_BUILD,
        codec::A_BIS,
        codec::A_CHOOSE_PLAN,
        codec::A_VALIDATE_ESTATE,
        codec::A_PASS_ROUNDABOUT,
        codec::A_PASS_SURVEYOR,
        codec::A_PASS_ESTATE,
        codec::A_PASS_PARK,
        codec::A_PASS_POOL,
        codec::A_PASS_BIS,
        codec::A_PASS_PLAN,
        codec::A_RESHUFFLE_YES,
        codec::A_RESHUFFLE_NO,
    ] {
        out.push(offset as i64);
    }
    for &(i, j) in codec::EXPERT_PAIRS.iter() {
        out.push(i as i64);
        out.push(j as i64);
    }

    out
}

/// FNV-1a 64 over `signature_stream`, each value as 8 little-endian bytes.
///
/// FNV rather than SHA-256 deliberately: the crate's dependency list starts at
/// pyo3 only (§6), and this hash guards against a transcription slip, not an
/// adversary.
pub fn table_signature() -> u64 {
    let mut h = FNV_OFFSET;
    for value in signature_stream() {
        let word = value as u64;
        for shift in (0..64).step_by(8) {
            h ^= (word >> shift) & 0xFF;
            h = h.wrapping_mul(FNV_PRIME);
        }
    }
    h
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The value `python -m games.welcome_to.tables` prints.
    ///
    /// This is the whole point of M0-D: if the tables here drift from the
    /// Python ones, this test fails before any game is played.
    #[test]
    fn signature_matches_python() {
        assert_eq!(signature_stream().len(), 697);
        assert_eq!(table_signature(), 0x2145_f101_3edc_fa99);
    }
}
